### PR TITLE
[IMP] l10n_ar_edi_ux: boarding permission

### PR DESCRIPTION
**Pull request by pablohmontenegro**
Task: 29241

### CODE REVIEW

**Comment by roboadhoc** [comment]
[Pull request status dashboard](https://mergebot.adhoc.ar/ingadhoc/odoo-argentina-ee/pull/282).

File: l10n_ar_edi_ux/views/account_move_view.xml [hunk 1]
```diff
@@ -15,4 +15,14 @@
         </field>
     </record>
 
+    <record id="view_move_form" model="ir.ui.view">
+        <field name="name">account.move.afip.form</field>
+        <field name="model">account.move</field>
+        <field name="inherit_id" ref="l10n_ar_edi.view_move_form"/>
+        <field name="arch" type="xml">
+        <field name='l10n_ar_afip_xml_response' position="after">
+                <field name='l10n_ar_boarding_permission_ids' widget="many2many_tags" attrs="{'invisible': [('l10n_ar_afip_ws', '!=', 'wsfex')]}"/>
```
Comment (zaoral) [inline l10n_ar_edi_ux/views/account_move_view.xml]: please move field to Other Info tab, in the section named as AFIP.
Also check if you can create a permission from this view, this part is important please review it, if not we need to add extra options to make it work. Something like agregar al m2m_tags esto para que `options="{'no_create_edit': False, 'no_create': True}"` The Create and Edit option should be shong, the Create option iteself should not be shown

File: l10n_ar_edi_ux/models/account_move.py [hunk 2]
```diff
@@ -16,6 +16,7 @@ class AccountMove(models.Model):
         states={'draft': [('readonly', False)]},
         help='Set this field if it is you are reporting debit/credit note and have not related invoice.'
         ' IMPORTANT: This is only applies on "Electronic Invoice - Web Service"')
+    l10n_ar_boarding_permission_ids = fields.Many2many('l10n_ar.boarding_permission', help="Solo se envía esta información si la factura es de exportación y el 'Concepto AFIP' es 'Products / Definitive export of goods'")
```
Comment (zaoral) [inline l10n_ar_edi_ux/models/account_move.py]: Add string "Permiso de Embarque" so this name is properly shown on the views

File: l10n_ar_edi_ux/models/l10n_ar_boarding_permission.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from odoo import fields, models
+
+
+class L10nArBoardingPermission(models.Model):
+    _name = "l10n_ar.boarding_permission"
+    _description = 'Boarding Permission'
+    _rec_name = 'l10n_ar_boarding_permission_dst_country'
+
+    l10n_ar_boarding_permission_number = fields.Char(string="Permiso de embarque")
+    l10n_ar_boarding_permission_dst_country = fields.Many2one('res.country', string="País donde se aplica permiso de embarque")
```
Comment (zaoral) [inline l10n_ar_edi_ux/models/l10n_ar_boarding_permission.py]: not need to add  l10n_ar_boarding_ as a prefix to all fields.

The rule for the fields name is:
- if it is a new model, one that have l10n_ar prefix then the fields does not have to have l10n_ar preffix
- If we add a method to an existing model, a generic one (does not have any l10n_ar preffix in the model name= then we need to add the l10n_ar preffix.

Also:
1. the dst_country string should be the field help instead
2. make both fields required

```suggestion
    number = fields.Char(string="Permiso de embarque", required=True)
    dst_country = fields.Many2one('res.country', string="Pais Origen", help="País donde se aplica permiso de embarque", required=True)
```

File: l10n_ar_edi_ux/models/account_move.py [hunk 4]
```diff
@@ -88,3 +89,26 @@ def _post(self, soft=True):
...
+        res = super(AccountMove, self).wsfex_get_cae_request(last_id, client)
+        ArrayOfPermisions = client.get_type('ns0:ArrayOfPermiso')
+        if int(self.l10n_latam_document_type_id.code) == 19 and int(self.l10n_ar_afip_concept) == 1:
+            permisos = self._get_permissions()
+            permiso_existente = "S" if permisos else "N"
+        else:
+            permisos = []
+            permiso_existente = ""
+        res.update({'Permisos': ArrayOfPermisions(permisos) if permisos else None})
+        res.update({'Permiso_existente': permiso_existente})
```
Comment (zaoral) [inline l10n_ar_edi_ux/models/account_move.py]: not need the else clause.
The original method already has this resolved.

Only the If is enough to add the permissions info if exist and needed 

```suggestion
        if int(self.l10n_latam_document_type_id.code) == 19 and int(self.l10n_ar_afip_concept) == 1:
            permisos = self._get_permissions()
            res.update({'Permisos': ArrayOfPermisions(permisos) if permisos else None})
            res.update({'Permiso_existente': "S" if permisos else "N"})
```

File: l10n_ar_edi_ux/models/l10n_ar_boarding_permission.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from odoo import fields, models
+
+
+class L10nArBoardingPermission(models.Model):
+    _name = "l10n_ar.boarding_permission"
+    _description = 'Boarding Permission'
+    _rec_name = 'l10n_ar_boarding_permission_dst_country'
```
Comment (zaoral) [inline l10n_ar_edi_ux/models/l10n_ar_boarding_permission.py]: remove rec_name and instead define the display_name following this format "permisonum (Pais)" 
You can accomplish this by creating a name_get() and name_search() methods in this model.
This way we are able to see and search by contry and by permission number

**Review by zaoral — CHANGES_REQUESTED** [review]

File: l10n_ar_edi_ux/models/account_move.py [hunk 6]
```diff
@@ -88,3 +89,30 @@ def _post(self, soft=True):
...
+        return res
+
+    def check_valid_boarding_permission(self):
+        """ Este método sirve para verificar que los permisos de embarque ingresados en la factura de exportación son válidos. Recibe las credenciales de autenticación, cuit del usuario representado, código de despacho y país de destino y verifica la existencia en la base de datos aduanera. """
+        client, auth = self.company_id._l10n_ar_get_connection(self.journal_id.l10n_ar_afip_ws)._get_client()
+        for perm in self.l10n_ar_boarding_permission_ids:
+            request_data = {'Auth': auth, 'Id_permiso': perm.number, 'Dst_merc': int(perm.dst_country.l10n_ar_afip_code)}
+            response = client.service['FEXCheck_Permiso'](auth, request_data)
```
Comment (zaoral) [inline l10n_ar_edi_ux/models/account_move.py]:
```suggestion
            response = client.service['FEXCheck_Permiso'](auth, Id_permiso=perm.number, Dst_merc=int(perm.dst_country.l10n_ar_afip_code))
```

File: l10n_ar_edi_ux/models/account_move.py [hunk 7]
```diff
@@ -16,6 +16,7 @@ class AccountMove(models.Model):
         states={'draft': [('readonly', False)]},
         help='Set this field if it is you are reporting debit/credit note and have not related invoice.'
         ' IMPORTANT: This is only applies on "Electronic Invoice - Web Service"')
+    l10n_ar_boarding_permission_ids = fields.Many2many('l10n_ar.boarding_permission', string="Permiso de Embarque", check_company=True, domain="[('company_id', 'in', [company_id, False])]", ondelete='restrict', help="Solo se envía esta información si la factura es de exportación y el 'Concepto AFIP' es 'Products / Definitive export of goods'")
```
Comment (zaoral) [inline l10n_ar_edi_ux/models/account_move.py]: does we need  domain="[('company_id', 'in', [company_id, False])]",? 
I think we should not have any permission share withing companies.
Instead, please set company_id as required field in boarding model

File: l10n_ar_edi_ux/models/account_move.py [hunk 8]
```diff
@@ -88,3 +89,30 @@ def _post(self, soft=True):
...
+    @api.model
+    def wsfex_get_cae_request(self, last_id, client):
+        """ Set permiso de embarque to foreign invoice. """
+        res = super(AccountMove, self).wsfex_get_cae_request(last_id, client)
+        ArrayOfPermisions = client.get_type('ns0:ArrayOfPermiso')
+        if int(self.l10n_latam_document_type_id.code) == 19 and int(self.l10n_ar_afip_concept) == 1:
+            permisos = self._get_permissions()
+            permiso_existente = "S" if permisos else "N"
+        res.update({'Permisos': ArrayOfPermisions(permisos) if permisos else None})
+        res.update({'Permiso_existente': permiso_existente})
```
Comment (zaoral) [inline l10n_ar_edi_ux/models/account_move.py]:
```suggestion
        if int(self.l10n_latam_document_type_id.code) == 19 and int(self.l10n_ar_afip_concept) == 1:
            ArrayOfPermisions = client.get_type('ns0:ArrayOfPermiso')
            permisos = self._get_permissions()
            permiso_existente = "S" if permisos else "N"
            res.update({'Permisos': ArrayOfPermisions(permisos) if permisos else None})
            res.update({'Permiso_existente': permiso_existente})
```

File: l10n_ar_edi_ux/models/l10n_ar_boarding_permission.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
+from odoo import fields, models, api
+
+
+class L10nArBoardingPermission(models.Model):
+    _name = "l10n_ar.boarding_permission"
+    _description = 'Boarding Permission'
+
+    number = fields.Char(string="Permiso de embarque", required=True, size=16)
+    dst_country = fields.Many2one('res.country', string="País Origen", help="País donde se aplica permiso de embarque", required=True)
+    company_id = fields.Many2one('res.company', default=lambda self: self.env.company)
```
Comment (zaoral) [inline l10n_ar_edi_ux/models/l10n_ar_boarding_permission.py]:
```suggestion
    company_id = fields.Many2one('res.company', required=True, default=lambda self: self.env.company)
```

File: l10n_ar_edi_ux/views/account_move_view.xml [hunk 10]
```diff
@@ -11,6 +11,8 @@
                 <div>
                     <field name="l10n_ar_afip_asoc_period_start" class="oe_inline"/> to <field name="l10n_ar_afip_asoc_period_end" class="oe_inline"/>
                 </div>
+                <field name='l10n_ar_boarding_permission_ids' widget="many2many_tags" attrs="{'invisible': [('l10n_ar_afip_ws', '!=', 'wsfex')]}" options="{'no_create_edit': False}"/>
```
Comment (zaoral) [inline l10n_ar_edi_ux/views/account_move_view.xml]:
```suggestion
                <field name='l10n_ar_boarding_permission_ids' widget="many2many_tags" attrs="{'invisible': [('l10n_ar_afip_ws', '!=', 'wsfex')]}" options="{'no_create': True, 'no_create_edit': False}"/>
```

File: l10n_ar_edi_ux/views/l10n_ar_boarding_permission_view.xml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    </record>
+
+    <record id="l10n_ar_boarding_permission_form" model="ir.ui.view">
+        <field name="name">l10n_ar.boarding_permission.form</field>
+        <field name="model">l10n_ar.boarding_permission</field>
+        <field name="arch" type="xml">
+            <form>
+                <group>
+                    <field name="number"/>
+                    <field name="dst_country"/>
```
Comment (zaoral) [inline l10n_ar_edi_ux/views/l10n_ar_boarding_permission_view.xml:22]: add company field to view

**Review by zaoral — CHANGES_REQUESTED** [review]

File: l10n_ar_edi_ux/views/account_move_view.xml (hunk 10 above)
Comment (pablohmontenegro) [reply l10n_ar_edi_ux/views/account_move_view.xml]: Thanks! I already tried the change suggested but is not making the expected result, it shows both options (1: create, 2: create and edit)

Comment (pablohmontenegro) [reply l10n_ar_edi_ux/views/account_move_view.xml]: @zaoral

File: l10n_ar_edi_ux/models/account_move.py [hunk 12]
```diff
@@ -88,3 +89,29 @@ def _post(self, soft=True):
             raise UserError(msg)
 
         return super(AccountMove, self - purchase_ar_edi_invoices)._post(soft=soft)
+
+    def _get_permissions(self):
+        """ Get 'permiso de embarque' for foreign invoices. """
+        self.ensure_one()
+        res = []
```
Comment (pablohmontenegro) [inline l10n_ar_edi_ux/models/account_move.py:96]: acá llamar al botón

**Comment by zaoral** [comment]
@roboadhoc r+

**Comment by zaoral** [comment]
@roboadhoc override=ci/runbot-oba

we have a warning not related to this pr https://runbot.adhoc.com.ar/runbot/build/30421

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#296

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#296

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#296

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#296

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#296

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#296

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#296

**Pull request closed without merging**